### PR TITLE
Fix build on macOS & rework build setup 

### LINKED ISSUE #411: Build without X.
Last night, I build  `emacs-ng`  using [xbps-src](https://github.com/void-linux/void-packages/) with template given below :-
````
# Template file for 'emacs-ng'
pkgname=emacs-ng
version=0.0.e88654b
revision=1
archs="x86_64"
wrksrc=${pkgname}-${version}
build_style=gnu-configure
hostmakedepends="base-devel clang"
makedepends="zlib-devel fontconfig-devel libotf-devel freetype-devel jansson-devel librsvg-devel gnutls-devel ncurses-devel libxml2-devel giflib-devel"
depends="desktop-file-utils hicolor-icon-theme"
short_desc="A new approach to Emacs - Including TypeScript, Threading, Async I/O, and WebRender."
maintainer=" <>"
license="GPL-3.0-or-later"
homepage="https://emacs-ng.github.io/emacs-ng"
distfiles="https://github.com/emacs-ng/emacs-ng/archive/refs/tags/v${version}.tar.gz"
checksum=2b00eb78f9479e4b8227b53b98ba788e29197c9c953d46b2193c014d5758ceac

export RUSTUP_HOME="${XBPS_BUILDDIR}/rustup"
export CARGO_HOME="${XBPS_BUILDDIR}/cargo"


do_configure(){
    # Installing Rust Nightly Toolchain
    mkdir -p ${RUSTUP_HOME} ${CARGO_HOME}
    RUSTUP_TOOLCHAIN=$(cat rust-toolchain)
    curl https://sh.rustup.rs -sSf | sh -s -- --no-modify-path --profile minimal --default-toolchain ${RUSTUP_TOOLCHAIN} -y
}

do_build(){
    export PATH=$CARGO_HOME/bin:$PATH
    ./autogen.sh
    ./configure CFLAGS="-Wl,-rpath,shared,--disable-new-dtags -g -O2" \
		LDFLAGS="-lxcb-xfixes -lxcb-shape -lxcb-render" \
		--prefix=/usr --enable-link-time-optimization\
		--with-json --with-modules --with-compress-install \
		--with-threads --with-included-regex --with-zlib \
		--with-native-compilation --with-webrender --with-otf \
		--with-sound --without-toolkit-scroll-bars --without-mailutils \
		--without-x-toolkit --without-gsettings --without-libsystemd \
		--without-makeinfo --without-x 

    make -j4 NATIVE_FULL_AOT=1 CARGO_FLAGS="-vv" RUSTFLAGS="-Zunstable-options --cfg MARKER_DEBUG"make CARGO_FLAGS="-vv" RUSTFLAGS="-Zunstable-options --cfg MARKER_DEBUG"
}

do_install(){
    export PATH=$CARGO_HOME/bin:$PATH
    make NATIVE_FULL_AOT=1 CARGO_FLAGS="-vv" RUSTFLAGS="-Zunstable-options --cfg MARKER_DEBUG"make CARGO_FLAGS="-vv" RUSTFLAGS="-Zunstable-options --cfg MARKER_DEBUG" DESTDIR=${PKGDESTDIR} install-strip

    rm -f ${PKGDESTDIR}/usr/bin/ctags
    rm -f ${PKGDESTDIR}/usr/bin/etags
    rm -f ${DESTDIR}/usr/share/man/man1/ctags.1*
    rm -f ${DESTDIR}/usr/share/man/man1/etags.1*
    rm -f ${DESTDIR}/usr/lib/systemd/user/emacs.service
}
````
I tried to build it without any X dependency as I am using Wayland and want to use Webrender.
Compilation went successful but I noticed something weird.

When I tried to open  `emacs` ,it didn't rendered fonts :

20211106_11h33m05s_grim

Whereas running  `emacs -q`  did it properly :

20211106_11h30m42s_grim
 
````
Configured for 'x86_64-pc-linux-gnu'.

  Where should the build process find the source code?    .
  What compiler should emacs be built with?               cc -Wl,-rpath,shared,--disable-new-dtags -g -O2 -flto=4 -ffat-lto-objects
  Should Emacs use the GNU version of malloc?             no
    (The GNU allocators don't work with this system configuration.)
  Should Emacs use a relocating allocator for buffers?    no
  Should Emacs use mmap(2) for buffer allocation?         no
  What window system should Emacs use?                    webrender
  What toolkit should Emacs use?                          none
  What libraries will lib-remacs link against?            -ldl -lm -lrt -lstdc++
  Where do we find X Windows header files?                NONE
  Where do we find X Windows libraries?                   NONE
  Does Emacs use -lXaw3d?                                 no
  Does Emacs use -lXpm?                                   no
  Does Emacs use -ljpeg?                                  no
  Does Emacs use -ltiff?                                  no
  Does Emacs use a gif library?                           no 
  Does Emacs use a png library?                           no 
  Does Emacs use -lrsvg-2?                                no
  Does Emacs use cairo?                                   no
  Does Emacs use -llcms2?                                 no
  Does Emacs use imagemagick?                             no
  Does Emacs use native APIs for images?                  yes (webrender)
  Does Emacs support sound?                               yes
  Does Emacs use -lgpm?                                   no
  Does Emacs use -ldbus?                                  yes
  Does Emacs use -lgconf?                                 no
  Does Emacs use GSettings?                               no
  Does Emacs use a file notification library?             yes -lglibc (inotify)
  Does Emacs use access control lists?                    no
  Does Emacs use -lselinux?                               no
  Does Emacs use -lgnutls?                                yes
  Does Emacs use -lxml2?                                  yes
  Does Emacs use -lfreetype?                              no
  Does Emacs use HarfBuzz?                                no
  Does Emacs use -lm17n-flt?                              no
  Does Emacs use -lotf?                                   no
  Does Emacs use -lxft?                                   no
  Does Emacs use -lsystemd?                               no
  Does Emacs use -ljansson?                               yes
  Does Emacs use the GMP library?                         yes
  Does Emacs directly use zlib?                           yes
  Does Emacs have dynamic modules support?                yes
  Does Emacs use toolkit scroll bars?                     no
  Does Emacs support Xwidgets?                            no
  Does Emacs have threading support in lisp?              yes
  Does Emacs support the portable dumper?                 yes
  Does Emacs support legacy unexec dumping?               no
  Which dumping strategy does Emacs use?                  pdumper
  Does Emacs have native lisp compiler?                   yes
  Does Emacs use webrender?                               yes
  Does Emacs include JavaScript?                          yes
````
This is what I have received during `Compilation`.

Is there something I missed while compiling or is this a issue with Webrender ??

### PR DESCRIPTION
**Pull request by declantsien**
Changes in this MR:
- disable js crate until it catches up
  `OPTION_DEFAULT_OFF` for javascript and commented out deno deps from js crate
  
- webrender: fix build on macOS
  * decouple colors as a crate
  * Replaced glutin with surfman
  * upgrade webrender
  * upgrade fontdb and drop font-kit

- build: Rework build setup
  * move Cargo workspace to top_srcdir
  * replaced ng-bindgen with emacs/build.rs
  * check rustc/cargo using configure, remove rustup checking
  * configure crates/webrender directly from configure.ac file

- Revert straight.el/use-package, since use-package is upstreamed in emaccs-29    

Should fix #411 #207 #242

### CODE REVIEW

**Comment by tiny911** [comment]
great work, but failed to compile on my macOS

OS 

> Darwin MacBook-Pro-3.local 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:52 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8112 arm6

Error 

> Caused by:
  process didn't exit successfully: `/Users/tiny/github/emacs-ng/target/release/build/emacs-8a8b3ba454479a4d/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=Cargo.toml
  cargo:rerun-if-changed=build.rs

  --- stderr
  ../../../src/gnutls.h:23:10: fatal error: 'gnutls/gnutls.h' file not found
  ../../../src/gnutls.h:23:10: fatal error: 'gnutls/gnutls.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', rust_src/crates/emacs/build.rs:303:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
make[2]: *** [../target/release/libemacsng.d] Error 101
make[1]: *** [src] Error 2

File: Cargo.toml.in
Comment (brotzeit) [inline Cargo.toml.in:84]: I can't see this repo, maybe it's private ?

**Comment by declantsien** [comment]
brotzeit ***@***.***> writes:

> +
> +[patch.crates-io.fontdb]
> +git = "https://github.com/declantsien/fontdb.git"
>
> I can't see this repo, maybe it's private ?
>

I accidently deleted my fork after upstream made the corresponding changes.
Currently I am exprementing against [emacs-29
branch](https://github.com/declantsien/emacs-ng/commits/emacs-29). Haven't
got time to back port changes to this PR.


Anyway, I restored the repo
https://github.com/declantsien/fontdb/tree/emacs-wr
Sorry about the inconvenience

**Comment by declantsien** [comment]
"xc.zhang" ***@***.***> writes:

>
> Error 
>
>> Caused by:
>   process didn't exit successfully: `/Users/tiny/github/emacs-ng/target/release/build/emacs-8a8b3ba454479a4d/build-script-build` (exit status: 101)
>   --- stdout
>   cargo:rerun-if-changed=Cargo.toml
>   cargo:rerun-if-changed=build.rs
>
>   --- stderr
>   ../../../src/gnutls.h:23:10: fatal error: 'gnutls/gnutls.h' file not found
>   ../../../src/gnutls.h:23:10: fatal error: 'gnutls/gnutls.h' file not found, err: true
>   thread 'main' panicked at 'Unable to generate bindings: ()', rust_src/crates/emacs/build.rs:303:10
>   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> make[2]: *** [../target/release/libemacsng.d] Error 101
> make[1]: *** [src] Error 2
>

Are you testing code from my latest changes.
This PR haven't included this file `rust_src/crates/emacs/build.rs` ...

This is the way to add new CFLAGS when some header files are missing.
https://github.com/declantsien/emacs-ng/commit/5c9561434a6f6d8885b5824c62718597f7a56a5d

File: Cargo.toml.in
Comment (declantsien) [reply Cargo.toml.in]: PR updated.

**Comment by declantsien** [comment]
With the latest updates, I got gh all ci tests except nix-build passed.
Here's the tests run
https://github.com/declantsien/emacs-ng/actions/runs/3797866203/jobs/6459149954

**Comment by tiny911** [comment]
After add  @LIBXCB_CFLAGS@ according [](https://github.com/declantsien/emacs-ng/commit/5c9561434a6f6d8885b5824c62718597f7a56a5d), still missing xcb


>     ../lib/libgnu.a                       -lxml2    -lncurses         -L/opt/homebrew/Cellar/gnutls/3.7.8/lib -lgnutls   -L/opt/homebrew/Cellar/little-cms2/2.14/lib -llcms2   -lz   -L/opt/homebrew/Cellar/jansson/2.14/lib -ljansson   -lxcb -lxcb-xfixes -lxcb-shape -lxcb-render -ldl -lm -lresolv -lstdc++
ld: library not found for -lxcb
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [temacs] Error 1
make: *** [src] Error 2

**Comment by declantsien** [comment]
"xc.zhang" ***@***.***> writes:

>>     ../lib/libgnu.a                       -lxml2    -lncurses
>>     -L/opt/homebrew/Cellar/gnutls/3.7.8/lib -lgnutls
>>     -L/opt/homebrew/Cellar/little-cms2/2.14/lib -llcms2   -lz
>>     -L/opt/homebrew/Cellar/jansson/2.14/lib -ljansson   -lxcb
>>     -lxcb-xfixes -lxcb-shape -lxcb-render -ldl -lm -lresolv -lstdc++

Sorry to hear that this didn't work for you. If you just want to try
webrender macOS build. I got some prebuild binaries here:

https://github.com/declantsien/emacs-wr/releases

It's not usable though.

**Comment by tiny911** [comment]
Success to build with latest commit [https://github.com/declantsien/emacs-ng/commit/7c692e718273662ff78c6ce0e0315476c57613ce](url), thanks very mush.

**Comment by GTrunSec** [comment]
So cool, thanks

**Comment by harryfei** [comment]
what'

File: rust_src/crates/webrender/src/event.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/event.rs:17]: unused line

File: rust_src/crates/webrender/src/event_loop.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/event_loop.rs:207]: why is this rename required?

File: rust_src/crates/webrender/src/event_loop.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/event_loop.rs:250]: Could we give those timeouts better names?

File: rust_src/crates/webrender/src/font.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/font.rs:448]: we should delete this before merge

File: rust_src/crates/webrender/src/font_db.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/font_db.rs:75]: We need delete this before merge

File: rust_src/crates/webrender/src/input.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/input.rs:53]: clean this file before merge.

File: rust_src/crates/webrender/src/output.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/output.rs:131]: I believe Linux wayland platform need those codes to have a better UI startup experience.

File: rust_src/crates/webrender/src/output.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/output.rs:549]: It looks to me if we hard code it to 24. But we need to clean unneeded lines.

File: rust_src/crates/webrender/src/output.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/output.rs:716]: clean it

File: rust_src/crates/webrender/src/output.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/output.rs:644]: use rust drop trait here, the output will be deinit in `destory_frame` automatically.

https://doc.rust-lang.org/std/ops/trait.Drop.html

File: rust_src/crates/webrender/src/term.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/term.rs:848]: rust's ownership will call `drop` trait function of a object automatically for us.

File: rust_src/crates/webrender/src/wrterm.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/wrterm.rs:222]: does those bools cause problems to you?
I think `lisp_fn` converts them into Qt/Qnils automatically.

File: src/wrterm.c
Comment (harryfei) [inline src/wrterm.c:36]: Do we have problems to implements those code in rust?

**Review by harryfei — CHANGES_REQUESTED** [review]

File: rust_src/crates/webrender/src/output.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/output.rs:500]: I missed this line in previous review

File: rust_src/crates/webrender/src/font.rs
Comment (harryfei) [inline rust_src/crates/webrender/src/font.rs:179]: I missed this line in previous review

**Review by harryfei — CHANGES_REQUESTED** [review]

**Comment by harryfei** [comment]
Is nix-build error caused by this PR. If not, I can merge this PR after you resolved all requested changes.

**Comment by declantsien** [comment]
Harry ***@***.***> writes:

> Is nix-build error caused by this PR. If not, I can merge this PR after you resolved all requested changes.
>

Nah. I believe that the nix-build is broken and did not caused by this PR.

I think I have already resolved all requested changes.
Are there requests pending that I missed?

**Comment by declantsien** [comment]
@harryfei All resolved  I think.

File: rust_src/crates/js/Cargo.toml
Comment (harryfei) [inline rust_src/crates/js/Cargo.toml:17]: by mistake?

**Review by harryfei — CHANGES_REQUESTED** [review]

**Comment by 2e0byo** [comment]
Surfman still fails on my graphics hardware.  The problem is somewhere in surfman itself, and I've forked + upgraded everything and narrowed it down to surfman.  (The demo results in a blank window locally with recent (>27) winit; with old winit we die with an illegal opcode).  This serves me right for having an nvidia gpu, but it's all I've got.

@harryfei @declantsien would you be open to a PR moving the rendering api back to glutin?

Surfman has a much nicer api and I *really* wanted to get it running when I saw what @declantsien had done with it, but it's beyond me.  The problem is somewhere in the GL pipeline itself.

Alternatively, @declantsien has already started moving the frame api away from one backend in the `feature/tao` branch.  There's no reason the rendering logic has to be pinned so tightly to the rendering toolkit: should we factor out a generic surface api and then implement that for surfman/glutin?

**Comment by 2e0byo** [comment]
> Can you try the webrender examples here https://github.com/jdm/webrender/tree/surfman

I get multiple panics during the build---looks like macro semantics/the implementation of `derive!()` has changed.  Could probably get it building with an older rust, but it's also targetting an old winit and would need updating to run.  FWIW the upstream demos build but don't actually open a window.  Thus I think I have particularly problematic hardware.  (Servo also fails here; I spent a day failing to get anything to run except glutin/winit and in the end wrote the logic blindly.  It was nice when emacs actually started up.)

> I am also wondering what is the cause. Since we only use surfman/glutin to get a OpenGL context.

I think it's not actually getting a valid context at all.  When I've got the glutin context running I'll try to compare them.

> I made some changes to make surfman directly works with `raw-window-handle`. So it can be used with tao or even PGTK. I am we can do that with glutin too.
> We could add cargo features for surfman/glutin. Emacs webrender just need a OpenGL Context from underline window as far as I know. Instead rely on winit, we rely on raw_window_handle. I think that is all.

We also need a few abstractions for things like resizing windows, but there's no reason they can't be abstracted.  I'll give it a go shortly.  Regardless I'll stop spamming this issue.  Thanks!

**Comment by harryfei** [comment]
@declantsien I get error after the PR merged on Arch Linux. The emacs-ng on Linux is unusable now.

```
Font ‘Monospace’ is not defined
```

You delete default font family logic in  https://github.com/emacs-ng/emacs-ng/commit/c160b360e9aad1400d922fffbf0a7e506ba84722. It's needed on Linux platform.
Can you revert those changes or re-implement the same functionality in a new way?

**Comment by declantsien** [comment]
Harry ***@***.***> writes:

> @declantsien I get error after the PR merged on Arch Linux. The emacs-ng on Linux is unusable now.
>
> ```
> Font ‘Monospace’ is not defined
> ```
>
> You delete default font family logic in  https://github.com/emacs-ng/emacs-ng/commit/c160b360e9aad1400d922fffbf0a7e506ba84722. It's needed on Linux platform.
> Can you revert those changes or re-implement the same functionality in a new way?
>
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/emacs-ng/emacs-ng/pull/456#issuecomment-1426664867
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>

I fixed this issue once for all here
https://github.com/declantsien/emacs-ng/commit/04c8fe8b47199ba1287d7d915ca9c6d654b28fb1
Haven't got time to submit a PR yet.

Sorry for the inconvenience.

**Pull request merged**
